### ISSUE TITLE
How to input non-English characters?

### ISSUE THREAD

**Issue by whatacold** [qtile/qtile#1447]
I'm not sure if this is a qtile-alone issue, but how do you input non-English characters?

**Comment by whatacold** [comment]
@zordsdavini For example, how to input Chinese in a search engine box?

**Comment by pawamoy** [comment]
Well you need to set your keyboard layout to a chinese one? `setxkbmap CODE` where the code is zh_CN for mainland China, zh_SG for Singapore, zh_TW for Taiwan, or zh_HK for Hong Kong?

**Comment by whatacold** [comment]
Wow, you're very familiar with Chinese locales, thanks for the info.

But I'm not sure how to invoke the input method, e.g. I can switch/activate the input method with `mod+space` when I was using GNome.

And it seems there are a few frameworks regarding input method on X (e.g. [XIM](https://en.wikipedia.org/wiki/X_Input_Method)), which are supported with qtile?
I may need to take some time to investigate it.

**Comment by whatacold** [comment]
Hmm, it seems it's [IBus](https://en.wikipedia.org/wiki/Intelligent_Input_Bus) on GNOME after looking closer to it, and Rime, one of the most popular Chinese IM on Linux, also supports IBus.

**Comment by pawamoy** [comment]
Aaaaah, yes of course, `mod+space`, I understand now.

I think you should check these two built-in Qtile widgets: [KeyboardLayout](http://docs.qtile.org/en/latest/manual/ref/widgets.html#keyboardlayout) and [KeyboardKbdd](http://docs.qtile.org/en/latest/manual/ref/widgets.html#keyboardkbdd).

**Comment by pawamoy** [comment]
In fact, look at [this code snippet](https://github.com/qtile/qtile/issues/1372#issue-464427979):

```python
@lazy.function
def z_next_keyboard(qtile):
    keyboard_widget.cmd_next_keyboard()

keys = [
    ...
    Key(["mod4"], "space", z_next_keyboard),
    ...
]

keyboard_widget = widget.KeyboardLayout(configured_keyboards=['en', 'zh'])
```

**Comment by zordsdavini** [comment]
don't forget to add `keyboard_widget` to bar, check [https://github.com/zordsdavini/dotfiles/blob/master/.config/qtile/config.py](my config) where I use it

**Comment by whatacold** [comment]
Thank you both for pointing out this widget, I never thought `KeyboardLayout` was related to input method, anyway, it sounds like a layout for windows :)

I initialize a `keyboard_widget = widget.KeyboardLayout(configured_keyboards=['en', 'zh'])` in the config, but nothing happens when I click it on the bar, and qtile complains like this (`setxkbmap` does exist on my OS):

    2019-09-17 22:28:19,145 ERROR libqtile keyboardlayout.py:keyboard():L116 Can not change the keyboard layout: Command '['setxkbmap', 'en']' returned non-zero exit status 251.

What was wrong? My locale is `en_US.UTF-8` (don't know if it's related). I didn't manage to find a way to list all possible layouts for `setxkbmap`, is there any way?

**Comment by whatacold** [comment]
> `setxkbmap CODE` where the code is zh_CN for mainland China, zh_SG for Singapore, zh_TW for Taiwan, or zh_HK for Hong Kong?

Well, `setxkbmap CODE` exits with code 251 with all of these four layouts :(

**Comment by zordsdavini** [comment]
What system do you use?

**Comment by pawamoy** [comment]
For US english it's `us`, not `en`, my bad. It's `gb` for GB english.

Run `ls /usr/share/X11/xkb/symbols` to see the available codes. I guess mainland china here is `cn`. I also have `tw` for Taiwan (not sure though), but no `sg` or `hk`.

You might also need to install additional software to actually be able to type chinese characters, as you said (`ibus-pinyin` or other "pinyin" software).

But yeah, not related to Qtile at all :sweat_smile:

**Comment by zordsdavini** [comment]
I read about Chinese input. Quite interesting stuff :)
To finish about KeyboardLayout widget I need to say:
- it use setxkbdmap command. If you don't have it - install :) To check is it working `setxkbmap  -layout us,lt,cn -option grp:alt_shift_toggle,compose:menu,grp_led:scroll`. This will add English,Lithuanian and Chinese input possibility, plus switch by ALT+SHIFT and compose key by Menu button, plus led is lighting if selected alternative input.
- for Chinese this is not enough, you need add awesome apps and deamons. This is because Chinese writing system is awesome :) So, to [read more](https://wiki.archlinux.org/index.php/IBus)
- our widget will not display anything what is set in console with command above. This is quite silly but it works. I think it is time to add 'options' option, because composite key for me is needfull

**Comment by whatacold** [comment]
@pawamoy @zordsdavini ,Thank you both for detailed replies.

I just realised what keyboard layouts were when I looked up WikiPedia article days before. I must say that most of keyboards use US layout in China, I haven't seen a keyboard with a non-US layout yet. So at least for me, I don't have to use `setxkbmap` and `KeyboardLayout` widget :)

Following the ArchWiki link above, it works by just starting the `ibus-daemon` by executing `ibus-daemon -drx`, it will display a panel on the system tray, so `Systray` widget must be put on the bar.

Here is my snippet to bring it up when logging:

    screens = [
        Screen(
            bottom=bar.Bar(
                [
                    ...
                    widget.Systray(),
                    ...
                ],
            ),
        ),
    ]
    
    def main(qtile):
        subprocess.run(["ibus-daemon", "-drx"])

And for GNOME, I finally find out it's brought up by `gnome-shell` on [windowsManager.js]() and then [ibusManager.js](), so the process is started with `ibus-daemon --xim --panel disable`. The reason for no panel needed here, I guess, is that gnome-shell provides its own "panel" for it.
